### ISSUE TITLE
Improvement: Retry undelivered notifications

### ISSUE THREAD

**Issue by pulkitsinghal** [pulkitsinghal/sample-notifier-service#1]
1. Detect if a notification cannot be delivered
1. Place the notification in a cache or queue or DB so that a delivery can be retried at a later time
    * If a browser reconnects with a different socketId then redelivery is not possible
    * Therefore, username based grouping of socketIds would make it much easier to assure that a notification is delivered, reliably.
        * It would also make it simpler to determine when a previously stored notification should be delivered.
            * For example: whenever a new sessionId is added for a username, it can be considered an event at which we should attempt to redeliver any and all pending notifications
                ```
                # find and broadcast any pending notifications when a new sockedId joins a username grouping
                {usernameA: [deadSocketId_A, mobileDeviceSocketId_B, NEW_SocketId_C]}
                {usernameB: [mobileDeviceSocketId_D, desktopSocketId_E]}
                ```

**Comment by pulkitsinghal** [comment]
@aquid - I have documented our whiteboard discussion here.

**Comment by pulkitsinghal** [comment]
Hello @yogesh8177, did you already accomplish all the key requirements from this discussion when you wrote https://github.com/ShoppinPal/notifier-service?

cc @aquid

**Comment by yogesh8177** [comment]
Yes, undelivered notifications are stored in queue (mongodb). Also when client reconnects, it can get those undelivered notifications!
